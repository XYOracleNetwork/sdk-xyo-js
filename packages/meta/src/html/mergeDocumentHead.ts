import { load } from 'cheerio'

export const mergeDocumentHead = (destination: string, source: string) => {
  const $destination = load(destination)
  const $source = load(source)

  // For each child node of the source head
  $source('head')
    .children()
    .each((_, element) => {
      const el = $destination(element)

      // Special case for meta tags: We want to match them by the name attribute
      if (el[0].tagName === 'meta') {
        const property = el.attr('property')
        if (property) {
          const match = $destination(`head meta[property="${property}"]`)

          // If it exists, replace it, otherwise append it
          if (match.length > 0) {
            match.replaceWith(el)
            return
          } else {
            $destination('head').append(el)
          }
        }
        // else {
        //   // For all other elements, just check if the same element exists in the first HTML string
        //   const match = $destination(el[0].tagName)

        //   // If it exists, replace it, otherwise append it
        //   if (match.length > 0) {
        //     match.replaceWith(el)
        //   } else {
        //     $destination('head').append(el)
        //   }
        // }
      }
    })

  // Return the merged HTML
  return $destination.html()
}
