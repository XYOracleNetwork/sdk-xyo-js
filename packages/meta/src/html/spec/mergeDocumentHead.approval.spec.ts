import { readFile, writeFile } from 'node:fs/promises'
import { join } from 'node:path'

import { mergeDocumentHead } from '../mergeDocumentHead'

const writeDebugFile = false

describe('mergeDocumentHead', () => {
  let destination: string
  let source: string
  beforeAll(async () => {
    destination = await readFile(join(__dirname, 'destination.html'), { encoding: 'utf8' })
    source = await readFile(join(__dirname, 'source.html'), { encoding: 'utf8' })
  })
  it('with real documents', async () => {
    const result = mergeDocumentHead(destination, source)
    expect(result).toMatchSnapshot()
    if (writeDebugFile) await writeFile(join(__dirname, 'result.html'), result)
  })
})
