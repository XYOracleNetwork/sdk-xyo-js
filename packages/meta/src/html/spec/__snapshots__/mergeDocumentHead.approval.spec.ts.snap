// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeDocumentHead with real documents 1`] = `
"<!DOCTYPE html><html lang="en"><head>
  <meta charset="utf-8">
  <link rel="icon" href="/favicon.ico">
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta name="description" content="Your Data, Forever Long.">
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="">
  <link href="https://fonts.googleapis.com/css2?family=Roboto+Flex:wght@300;400;500;600;800&amp;display=swap" rel="stylesheet">
  <link rel="apple-touch-icon" href="/foreventory-logo-icon-black.png">
  <link rel="manifest" href="/manifest.json">
  <title>Foreventory</title>
  <script defer="defer" src="/static/js/main.297e104a.js"></script>
<meta content="website" property="og:type" data-react-helmet="true"><meta property="og:image" content="/netflix/insights/6f088236519e8e0c9ab9294c2e59fb2b8b434091a7559401ae860ff445fb1355/share/1600/900"><meta property="og:image:height" content="900"><meta property="og:image:secure_url" content="/netflix/insights/6f088236519e8e0c9ab9294c2e59fb2b8b434091a7559401ae860ff445fb1355/share/1600/900"><meta property="og:image:type" content="image/png"><meta property="og:image:url" content="/netflix/insights/6f088236519e8e0c9ab9294c2e59fb2b8b434091a7559401ae860ff445fb1355/share/1600/900"><meta property="og:image:width" content="1600"><meta property="twitter:image" content="/netflix/insights/6f088236519e8e0c9ab9294c2e59fb2b8b434091a7559401ae860ff445fb1355/share/1600/900"></head>

<body><noscript>You need to enable JavaScript to run this app.</noscript>
  <div id="root"></div>


</body></html>"
`;
